README
=== CHANGES ===

v2.1 - Stamps out a Y2K bug and now uses GNU autoconf.


=== SUMMARY ===

"xtail" watches the growth of files.  It's like running a "tail -f"
on a bunch of files at once.  The syntax is:

	xtail name ...

You can specify both filenames and directories on the command line.
If you specify a directory, it watches all the files in that
directory.  It will notice when new files are created (and start
watching them) or when old files are deleted (and stop watching
them).

I originally wrote this program so I could run:

	xtail /usr/spool/uucp/.Log/*

Even though I don't use "uucp" any more, I still use this utility
a lot.  These days, my favorite is:

	xtail /var/log/* /usenet/lib/*log


=== INSTRUCTIONS ===

	- xtail version 2 requires an ANSI compiler such as gcc.
	- Run:  sh configure
	- If you'd like, review the configuration definitions in xtail.h
	- Run:  make
	- If you'd like, as root run:  make install


=== HISTORICAL NOTES ===

This program is an oldie but goodie.  It was posted to comp.sources.misc
in July 1989 (see ftp.uu.net:/usenet/comp.sources.misc/volume7/xtail.Z).
I remember posting an even earlier version to alt.sources.  It has been
published in the O'Reilly & Associates "Unix Power Tools" collection
(book and CD-ROM).

Over the years, some fly-by-night organizations (such as the MIT X
Consortium and SGI) have tried to steal the "xtail" name.  Don't be
fooled!  Insist on the original.

The 1989 release credited David Dykstra for his contributions.

Chip Rosenthal
<chip@unicom.com>

$Id: README,v 2.4 2000/06/04 05:15:35 chip Exp $


debian/changelog
xtail (2.1-8) unstable; urgency=medium

  * debian/control:
      - Bumped debhelper-compat to 13
  * New maintainer (Closes: #862650)
  * fix compile warning: implicit declaration of function ‘strcmp’
  * fix compile warning: format ‘%d’ expects argument of type ‘int’

 -- xiao sheng wen <atzlinux@sina.com>  Tue, 05 May 2020 18:16:47 +0800

xtail (2.1-7) unstable; urgency=medium

  * QA upload.
  * Using new DH level format. Consequently:
      - debian/compat: removed.
      - debian/control: changed from 'debhelper' to 'debhelper-compat' in
        Build-Depends field and bumped level to 12.
  * debian/control:
      - Added 'Rules-Requires-Root: no' to source stanza.
      - Bumped Standards-Version to 4.5.0.
      - Created VCS fields.
      - Reorganized the long description.
  * debian/copyright:
      - Reverted packaging licensing to BSD-3-Clause. (Closes: #956140)
      - Updated packaging copyright years.
  * debian/patches: added a numeric prefix for all patches.
  * debian/rules:
      - Added DEB_BUILD_MAINT_OPTIONS to improve the GCC hardening.
      - Removed trash.
  * debian/salsa-ci.yml: added to provide CI tests for Salsa.
  * debian/tests/control: created to perform a trivial CI test.
  * debian/watch: bumped version to 4.
  * debian/xtail.dirs: renamed to debian/dirs.

 -- Joao Eriberto Mota Filho <eriberto@debian.org>  Tue, 07 Apr 2020 00:39:15 -0300

xtail (2.1-6) unstable; urgency=medium

  * QA upload.
  * Source format 3.0 (quilt)
  * Bumped compat level to 10
  * Converted d/rules to dh format
  * Add ISO Date patch from Jari Aalto <jari.aalto@cante.net> (Closes: #562903)
  * Set maintainer to QA Group
  * Added homepage to d/control
  * Converted d/copyright to machine-readable format
  * Made URLs use HTTPS where possible
  * Standards version 4.0.0

 -- David William Richmond Davies-Jones <david@exultantmonkey.co.uk>  Wed, 19 Jul 2017 15:32:38 +0100

xtail (2.1-5.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Update DH level to 9 to avoid a FTBFS. (Closes: #800269)
  * debian/compat: created.
  * debian/control:
      - Added the ${misc:Depends} variable to provide the
        right install dependencies.
      - Bumped Standards-Version to 3.9.6.
  * debian/rules: disabled the DH_COMPAT line.
  * debian/watch: updated. (Closes: #551417)

 -- Joao Eriberto Mota Filho <eriberto@debian.org>  Thu, 10 Dec 2015 17:54:42 -0200

xtail (2.1-5) unstable; urgency=low

  * Note in debian/control that this isn't an X11 program (closes:
    #426350).

 -- Roderick Schertler <roderick@argon.org>  Tue, 24 Jun 2008 11:25:57 -0400

xtail (2.1-4) unstable; urgency=low

  * Add do-nothing binary/indep target (closes: #395763).
  * Oops, files in directories given on the command line were output in
    full due to a bug in code I'd introduced (closes: #368897)

 -- Roderick Schertler <roderick@argon.org>  Sat, 18 Nov 2006 21:39:17 -0500

xtail (2.1-3) unstable; urgency=low

  * Drop /usr/doc symlink by rebuilding with newer debhelper (closes:
    #359611).
  * Note in the copyright file that that Chip Rosental really is the
    author (closes: #302920).
  * xtail.1:  Change embedded section from "1L" to "1".

 -- Roderick Schertler <roderick@argon.org>  Thu,  6 Apr 2006 10:25:06 -0400

xtail (2.1-2) unstable; urgency=low

  * Handle files which are renamed properly.

 -- Roderick Schertler <roderick@argon.org>  Fri,  7 Dec 2001 09:16:54 -0500

xtail (2.1-1) unstable; urgency=low

  * Initial version (closes: #111309).

 -- Roderick Schertler <roderick@argon.org>  Thu, 13 Sep 2001 17:23:10 -0400

$Id: changelog,v 1.5 2008-06-24 15:26:26 roderick Exp $


xtail.c
/* $Id: xtail.c,v 2.5 2000/06/04 09:09:03 chip Exp $ */

#include "config.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <signal.h>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#ifdef HAVE_STDLIB_H
# include <stdlib.h>
#endif
#define  INTERN
#include "xtail.h"

int sigcaught = 0;
RETSIGTYPE sigcatcher(int sig);


int main(int argc, char *argv[])
{
    int open_files_only, already_open, iteration, i;
    struct entry_descrip *entryp;
    struct stat sbuf;

    /* 
     * Initialize.
     */
    Progname = basename(argv[0]);
    List_file = new_entry_list(ENTRY_LIST_CHUNK);
    List_dir = new_entry_list(ENTRY_LIST_CHUNK);
    List_zap = new_entry_list(ENTRY_LIST_CHUNK);
    Sorted = FALSE;
    Reset_status = FALSE;
    Debug = FALSE;
    sigcatcher(0);


    /*
     * Place all of the entries onto lists.
     */
    for (i = 1 ; i < argc ; ++i)  {

	if (i == 1 && strcmp(argv[i], "-D") == 0) {
	    Debug = TRUE;
	    continue;
	}

	/*
	 * Temporarily throw this entry onto the end of the zapped list.
	 */
	entryp = new_entry(List_zap, argv[i]);

	/*
	 * Stat the file and get it to its proper place.
	 */
	switch (stat_entry(List_zap, last_entry(List_zap), &sbuf)) {

	case ENTRY_FILE:		/* move entry to file list	*/
	    move_entry(List_file, List_zap, last_entry(List_zap));
	    entryp->size = sbuf.st_size;
	    entryp->mtime = sbuf.st_mtime;
	    break;

	case ENTRY_DIR:			/* move entry to dir list	*/
	    move_entry(List_dir, List_zap, last_entry(List_zap));
	    entryp->size = sbuf.st_size;
	    entryp->mtime = sbuf.st_mtime;
	    if (scan_directory(entryp->name) != 0) {
		message(MSSG_OPEN, entryp);
		rmv_entry(List_dir, last_entry(List_dir));
	    }
	    break;

	case ENTRY_ZAP:			/* keep entry on zap list	*/
	    break;

	case ENTRY_SPECIAL:		/* entry is a special file	*/
	    message(MSSG_NOTAFIL, entryp);
	    rmv_entry(List_zap, last_entry(List_zap));
	    break;

	default:			/* stat error			*/
	    message(MSSG_STAT, entryp);
	    rmv_entry(List_zap, last_entry(List_zap));
	    break;

	}

    }

    /*
     * Make sure we are watching something reasonable.
     */
    if (List_file->num_entries == 0) {
	if (List_dir->num_entries == 0 && List_zap->num_entries == 0) {
	    (void) fprintf(stderr, "%s: no valid entries specified\n", Progname);
	    (void) exit(1);
	}
	(void) puts("\n*** warning - no files are being watched ***");
    }


    /*
     * From this point on we want to reset the status of an entry any
     * time we move it around to another list.
     */
    Reset_status = TRUE;


    /*
     * Force a check of everything first time through the loop.
     */
    iteration = CHECK_COUNT;


    /* 
     * Loop forever.
     */
    for (;;) {

	/*
	 * Once every CHECK_COUNT iterations check everything.
	 * All other times only look at the opened files.
	 */
	open_files_only = (++iteration < CHECK_COUNT);
	if (!open_files_only)
	    iteration = 0;


	/*
	 * Make sure that the most recently modified files are open.
	 */
	if (!Sorted)
	    fixup_open_files();


	/*
	 * Display what we are watching if a SIGINT was caught.
	 */
	if (sigcaught) {
	    show_status();
	    sigcatcher(0);
	}


	/*
	 * Go through all of the files looking for changes.
	 */
	Dprintf(stderr, ">>> checking files list (%s)\n",
	    (open_files_only ? "open files only" : "all files"));
	for (i = 0 ; i < List_file->num_entries ; ++i) {

	    entryp = List_file->list[i];
	    already_open = (entryp->fd > 0) ;

	    /*
	     * Ignore closed files except every CHECK_COUNT iterations.
	     */
	    if (!already_open && open_files_only)
		continue;

	    /*
	     * Get the status of this file.
	     */
	    switch (stat_entry(List_file, i, &sbuf)) {
	    case ENTRY_FILE:		/* got status OK		*/
		break;
	    case ENTRY_DIR:		/* huh??? it's now a dir	*/
		move_entry(List_dir, List_file, i--);
		continue;
	    case ENTRY_ZAP:		/* entry has been deleted	*/
		message(MSSG_ZAPPED, entryp);
		move_entry(List_zap, List_file, i--);
		continue;
	    case ENTRY_SPECIAL:		/* entry is a special file	*/
		message(MSSG_NOTAFIL, entryp);
		rmv_entry(List_file, i--);
		continue;
	    default:			/* stat error			*/
		message(MSSG_STAT, entryp);
		rmv_entry(List_file, i--);
		continue;
	    }


	    /*
	     * See if an opened file has been deleted.
	     */
	    if (already_open && sbuf.st_nlink == 0) {
		message(MSSG_ZAPPED, entryp);
		move_entry(List_zap, List_file, i--);
		continue;
	    }

	    /*
	     * If nothing has changed then continue on.
	     */
	    if (entryp->size==sbuf.st_size && entryp->mtime==sbuf.st_mtime)
		continue;

	    /*
	     * If the file isn't already open, then do so.
	     *   Note -- it is important that we call "fixup_open_files()"
	     *   at the end of the loop to make sure too many files don't
	     *   stay opened.
	     */
	    if (!already_open && open_entry(List_file, i) != 0) {
		--i;
		continue;
	    }

	    /*
	     * See if the file has been truncated.
	     */
	    if (sbuf.st_size < entryp->size) {
		message(MSSG_TRUNC, entryp);
		entryp->size = 0;
	    }

	    /*
	     * Seek to where the changes begin.
	     */
	    if (lseek(entryp->fd, entryp->size, 0) < 0) {
		message(MSSG_SEEK, entryp);
		rmv_entry(List_file, i--);
		continue;
	    }

	    /*
	     * Dump the recently added info.
	     */
	    {
		int nb;
    		static char buf[BUFSIZ];
		message(MSSG_BANNER, entryp);
		while ((nb = read(entryp->fd, buf, sizeof(buf))) > 0) {
		    (void) fwrite(buf, sizeof(char), (unsigned) nb, stdout);
		    entryp->size += nb;
		}
		if (nb < 0) {
		    message(MSSG_READ, entryp);
		    rmv_entry(List_file, i--);
		    continue;
		}
	    }

	    /*
	     * Update the modification time.
	     */
	    entryp->mtime = sbuf.st_mtime;

	    /*
	     * Since we've changed the mtime, the list might no longer be
	     * sorted.  However if this entry is already at the top of the
	     * list then it's OK.
	     */
	    if (i != 0)
		Sorted = FALSE;

	    /*
	     * If we've just opened the file then force a resort now to
	     * prevent too many files from being opened.
	     */
	    if (!already_open)
		fixup_open_files();

	}


	/*
	 * Go through list of nonexistent entries to see if any have appeared.
	 *   This is done only once every CHECK_COUNT iterations.
	 */
	if (!open_files_only) {
	    Dprintf(stderr, ">>> checking zapped list\n");
	    for (i = 0 ; i < List_zap->num_entries ; ++i) {
		entryp = List_zap->list[i];
		switch (stat_entry(List_zap, i, &sbuf)) {
		case ENTRY_FILE:	/* entry has appeared as a file	*/
		    message(MSSG_CREATED, entryp);
		    move_entry(List_file, List_zap, i--);
		    break;
		case ENTRY_DIR:		/* entry has appeared as a dir	*/
		    message(MSSG_CREATED, entryp);
		    move_entry(List_dir, List_zap, i--);
		    break;
		case ENTRY_ZAP:		/* entry still doesn't exist	*/
		    break;
		case ENTRY_SPECIAL:	/* entry is a special file	*/
		    message(MSSG_NOTAFIL, entryp);
	    	    rmv_entry(List_zap, i--);
		    break;
		default:		/* error - entry removed	*/
	    	    message(MSSG_STAT, entryp);
	    	    rmv_entry(List_zap, i--);
		    break;
		}
	    }
	}


	/*
	 * Go through the list of dirs to see if any new files were created.
	 *   This is done only once every CHECK_COUNT iterations.
	 */
	if (!open_files_only) {
	    Dprintf(stderr, ">>> checking directory list\n");
	    for (i = 0 ; !open_files_only && i < List_dir->num_entries ; ++i) {
		entryp = List_dir->list[i];
		switch (stat_entry(List_dir, i, &sbuf)) {
		case ENTRY_DIR:		/* got status OK		*/
		    break;
		case ENTRY_FILE:	/* huh??? it's now a reg file	*/
		    move_entry(List_file, List_dir, i--);
		    continue;
		case ENTRY_ZAP:		/* entry has been deleted	*/
	    	    message(MSSG_ZAPPED, entryp);
	    	    move_entry(List_zap, List_dir, i--);
		    continue;
		case ENTRY_SPECIAL:	/* entry is a special file	*/
		    message(MSSG_NOTAFIL, entryp);
		    rmv_entry(List_dir, i--);
		    continue;
		default:		/* stat error			*/
		    message(MSSG_STAT, entryp);
		    rmv_entry(List_dir, i--);
		    continue;
		}
		if (entryp->mtime == sbuf.st_mtime)
		    continue;
		if (scan_directory(entryp->name) != 0) {
		    message(MSSG_OPEN, entryp);
		    rmv_entry(List_dir, i--);
		}
		entryp->mtime = sbuf.st_mtime;
	    }
	}


	/*
	 * End of checking loop.
	 */
	{
	    extern unsigned sleep();
	    (void) fflush(stdout);
	    (void) sleep(SLEEP_TIME);
	}

    }

    /*NOTREACHED*/

}


RETSIGTYPE sigcatcher(int sig)
{
    if (sig == SIGQUIT)
	(void) exit(0);
    sigcaught = sig;
#ifdef STATUS_ENAB
    (void) signal(SIGINT, sigcatcher);
    (void) signal(SIGQUIT, sigcatcher);
#endif
}



xtail.h
/* $Id: xtail.h,v 2.7 2000/06/04 19:02:23 chip Exp $ */


/*****************************************************************************
 *
 * Start of Site-Specific Customizations
 *
 *****************************************************************************/

/*
 * STATUS_ENAB - If defined, a SIGINT causes a summary of the opened files to
 * be displayed, and a SIGQUIT terminates the program.  If not defined,
 * these signals act normally.
 */
#define STATUS_ENAB

/*
 * SLEEP_TIME - Number of seconds between iterations of the checking loop.
 */
#define SLEEP_TIME 1

/*
 * CHECK_COUNT - Recently modified files are checked for changes every time
 * through the checking loop.  We only go looking for changes to
 * not-so-recently modified stuff only once ever CHECK_COUNT iterations
 * through the loop.
 */
#define CHECK_COUNT 5

/*
 * MAX_OPEN - This number of most recently changed files is kept open, and
 * they are checked every iteration through the checking loop.  Keeping these
 * files open improves the performance because we can use "fstat()" rather
 * than "stat()".  Keeping too many files open may overflow your open file
 * table, and will reduce performance by checking more files more frequently.
 */
#define MAX_OPEN 8

/*
 * MAX_ENTRIES - This is *BOGUS*  I should get rid of this.
 */
#define MAX_ENTRIES 512



/*****************************************************************************
 *
 * End of Site-Specific Customizations
 *
 *****************************************************************************/


#define TRUE 1
#define FALSE 0

#define Dprintf if ( !Debug ) ; else (void) fprintf


/*
 * Codes returned by the "stat_entry()" procedure.
 */
#define ENTRY_ERROR	0	/* stat error or permissions error	*/
#define ENTRY_SPECIAL	1	/* entry is a special file		*/
#define ENTRY_FILE	2	/* entry is a regular file		*/
#define ENTRY_DIR	3	/* entry is a directory			*/
#define ENTRY_ZAP	4	/* specified entry doesn't exist	*/


/*
 * Diagnostic message codes.
 *   The ordering of codes must correspond to the "mssg_list[]" defined below.
 */
#define MSSG_NONE	0	/* no message - just reset header	*/
#define MSSG_BANNER	1	/* display banner for file output	*/
#define MSSG_CREATED	2	/* file has been created		*/
#define MSSG_ZAPPED	3	/* file has been deleted		*/
#define MSSG_TRUNC	4	/* file has been truncated		*/
#define MSSG_NOTAFIL	5	/* error - not a regular file or dir	*/
#define MSSG_STAT	6	/* error - stat() failed		*/
#define MSSG_OPEN	7	/* error - open() failed		*/
#define MSSG_SEEK	8	/* error - lseek() failed		*/
#define MSSG_READ	9	/* error - read() failed		*/
#define MSSG_UNKNOWN	10	/* unknown error - must be last in list */


#ifdef INTERN
#   define EXTERN
#else
#   define EXTERN extern
#endif


/*
 * Each item we are watching is stored in a (struct entry_descrip).  These
 * entries are placed in lists, which are managed as (struct entry_list).
 *
 * There are three lists maintained:
 *
 * List_file	All of the regular files we are watching.  We will try to
 *		keep the MAX_OPEN most recently modified files open, and
 *		they will be checked more frequently.
 *
 * List_dir	All of the directories we are watching.  If a file is created
 *		in one of these directories, we will add it to "List_file".
 *
 * List_zap	All the entries which don't exist.  When something appears
 *		under one of these names, the entry will be moved to either
 *		"List_file" or "List_dir", as appropriate.
 */

struct entry_descrip {
    char *name;		/* pathname to the entry			*/
    int fd;		/* opened fd, or <= 0 if not opened		*/
    long size;		/* size of entry last time checked		*/
    long mtime;		/* modification time last time checked		*/
};

struct entry_list {
    struct entry_descrip **list;
    int num_entries;	/* num entries stored in the list		*/
    int max_entries;	/* list allocated to hold this many entries	*/
    int chunk;		/* chunk for growin the list			*/
};

#define ENTRY_LIST_CHUNK 64
#define last_entry(L)	((L)->num_entries - 1)

/*
 * The lists of entries being watched.
 */
EXTERN struct entry_list *List_file;	/* regular files		*/
EXTERN struct entry_list *List_dir;	/* directories			*/
EXTERN struct entry_list *List_zap;	/* nonexistent entries		*/


/*
 * List sorting status.
 *   This flag indicates that "List_file" is sorted, and the right entries
 *   are open.  Anything which possibly effects this state (e.g. an entry
 *   is added to "List_file", the mtime of a file is changed, etc.) must set
 *   this flag FALSE.  We will periodically check this flag and call the
 *   "fixup_open_files()" procedure to resort and organize the list.
 */
EXTERN int Sorted;


/*
 * Entry status control flag.
 *   The procedures which manipulate entries will reset the status information
 *   if this flag is TRUE.  When initializing the lists we want this FALSE.
 *   For example, consider the file size.  When initializing we want to use
 *   the current file size, otherwise we would dump the file from the beginning.
 *   However, later when we notice things are created we want to reset the
 *   size to zero so that we do dump from the beginning.
 */
EXTERN int Reset_status;


EXTERN int Debug;
EXTERN char *Progname;


/*
 * Diagnostic messages produced by the "message()" procedure.
 *   The first "%s" is the entry name.  The second "%s" is the errno descrip.
 */
#ifdef INTERN
    char *mssg_list[] = {
	NULL,							/*MSSG_NONE   */
	"\n*** %s ***\n",					/*MSSG_BANNER */
	"\n*** '%s' has been created ***\n",			/*MSSG_CREATED*/
	"\n*** '%s' has been deleted ***\n",			/*MSSG_ZAPPED */
	"\n*** '%s' has been truncated - rewinding ***\n",	/*MSSG_TRUNC  */
	"\n*** error - '%s' not a file or dir - removed ***\n",	/*MSSG_NOTAFIL*/
	"\n*** error - couldn't stat '%s' (%s) - removed ***\n",/*MSSG_STAT   */
	"\n*** error - couldn't open '%s' (%s) - removed ***\n",/*MSSG_OPEN   */
	"\n*** error - couldn't seek '%s' (%s) - removed ***\n",/*MSSG_SEEK   */
	"\n*** error - couldn't read '%s' (%s) - removed ***\n",/*MSSG_READ   */
	"\n*** error - unknown error on file '%s' ***\n",	/*MSSG_UNKNOWN*/
    };
#else
    extern char *mssg_list[];
#endif


/*
 * Entry management procedures.
 */
struct entry_list *new_entry_list(int chunk);
struct entry_descrip *new_entry(struct entry_list *listp, const char *name);
void rmv_entry(struct entry_list *listp, int entryno);
void move_entry(struct entry_list *dst_listp, struct entry_list *src_listp,
	int src_entryno);
int stat_entry(struct entry_list *listp, int entryno, struct stat *sbuf);
int open_entry(struct entry_list *listp, int entryno);

/*
 * Miscellaneous procedures.
 */
int scan_directory(const char *dirname);
int ecmp(const void *p1, const void *p2);
void fixup_open_files(void);
void message(int sel, const struct entry_descrip *e);
void show_status(void);
char *quit_ch(void);
VOID *safe_malloc(size_t n);
VOID *safe_realloc(VOID *p, size_t n);
char *safe_strdup(const char *p);
char *basename(char *p);
#ifndef HAVE_DIFFTIME
double difftime(time_t t1, time_t t0)
#endif
#ifndef HAVE_STRERROR
char *strerror(int err);
#endif



entryfuncs.c
/* $Id: entryfuncs.c,v 2.6 2000/06/05 07:10:57 chip Exp $ */

#include "config.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#ifdef HAVE_STDLIB_H
# include <stdlib.h>
#endif

#include "xtail.h"

struct entry_list *new_entry_list(int chunk)
{
    struct entry_list *listp;
    listp = safe_malloc(sizeof(struct entry_list));
    listp->list = safe_malloc(chunk * sizeof(struct entry_descrip *));
    listp->num_entries = 0;
    listp->max_entries = chunk;
    listp->chunk = chunk;
    return listp;
}


static struct entry_descrip *E_append(struct entry_list *listp,
	struct entry_descrip *entryp)
{
    if (listp->num_entries >= listp->max_entries) {
	listp->max_entries += listp->chunk;
    	listp->list = safe_realloc(listp->list,
	    listp->max_entries * sizeof(struct entry_descrip *));
    }
    listp->list[listp->num_entries++] = entryp;
    Sorted = FALSE;
    return entryp;
}


static void E_remove(struct entry_list *listp, int entryno)
{
    while (++entryno < listp->num_entries)
	listp->list[entryno-1] = listp->list[entryno];
    --listp->num_entries;
    Sorted = FALSE;
}


static char *list_name(struct entry_list *listp) /* for debug output only */
{
    if (listp == List_file)	return "<file>";
    if (listp == List_dir)	return "<dir>";
    if (listp == List_zap)	return "<zap>";
    return "?unknown?";
}


/*
 * Create a new entry description and append it to a list.
 */
struct entry_descrip *new_entry(struct entry_list *listp, const char *name)
{
    struct entry_descrip *entryp;

    Dprintf(stderr, ">>> creating entry '%s' on %s list\n",
	name, list_name(listp));

    entryp = safe_malloc(sizeof(struct entry_descrip));
    entryp->name = safe_strdup(name);
    entryp->fd = 0;
    entryp->size =  0;
    entryp->mtime = 0;

    return E_append(listp,entryp);
}


/*
 * Remove an entry from a list and free up its space.
 */
void rmv_entry(struct entry_list *listp, int entryno)
{
    struct entry_descrip *entryp = listp->list[entryno];

    Dprintf(stderr, ">>> removing entry '%s' from %s list\n",
	listp->list[entryno]->name, list_name(listp));
    E_remove(listp,entryno);
    if (entryp->fd > 0)
	(void) close(entryp->fd);
    free((VOID *) entryp->name);
    free((VOID *) entryp);
}


/*
 * Move an entry from one list to another.
 *	In addition we close up the entry if appropriate.
 */
void move_entry(struct entry_list *dst_listp, struct entry_list *src_listp,
	int src_entryno)
{
    struct entry_descrip *entryp = src_listp->list[src_entryno];

    Dprintf(stderr, ">>> moving entry '%s' from %s list to %s list\n",
	src_listp->list[src_entryno]->name,
	list_name(src_listp), list_name(dst_listp));
    if (entryp->fd > 0) {
	(void) close(entryp->fd);
	entryp->fd = 0;
    }
    E_remove(src_listp,src_entryno);
    (void) E_append(dst_listp,entryp);
    if (Reset_status) {
	entryp->size = 0;
	entryp->mtime = 0;
    }
}


/*
 * Get the inode status for an entry.
 *	Returns code describing the status of the entry.
 */
int stat_entry(struct entry_list *listp, int entryno, struct stat *sbuf)
{
    int status;
    struct entry_descrip *entryp = listp->list[entryno];

    status = 
	(entryp->fd > 0 ? fstat(entryp->fd,sbuf) : stat(entryp->name,sbuf));

    if (status != 0)
	return (errno == ENOENT ? ENTRY_ZAP : ENTRY_ERROR);

    switch (sbuf->st_mode & S_IFMT) {
	case S_IFREG:	return ENTRY_FILE;
	case S_IFDIR:	return ENTRY_DIR;
	default:	return ENTRY_SPECIAL;
    }

    /*NOTREACHED*/
}


/*
 * Open an entry.
 *	Returns 0 if the open is successful, else returns errno.  In the case
 *	of an error, an appropriate diagnostic will be printed, and the entry
 *	will be moved or deleted as required.  If the entry is already opened,
 *	then no action will occur and 0 will be returned.
 */
int open_entry(struct entry_list *listp, int entryno)
{
    struct entry_descrip *entryp = listp->list[entryno];

    if (entryp->fd > 0)
	return 0;

    Dprintf(stderr, ">>> opening entry '%s' on %s list\n",
	listp->list[entryno]->name, list_name(listp));
    if ((entryp->fd = open(entryp->name, O_RDONLY)) > 0)
	return 0;

    if (errno == ENOENT) {
	message(MSSG_ZAPPED, entryp);
	move_entry(List_zap, listp, entryno);
    } else {
	message(MSSG_OPEN, entryp);
	rmv_entry(listp, entryno);
    }
    return -1;
}



